File: system/chre/util/pigweed/chre_channel_output.cc
/*
 * Copyright (C) 2022 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "chre/util/pigweed/chre_channel_output.h"

#include "chre/util/memory.h"
#include "chre/util/nanoapp/callbacks.h"

namespace chre {
namespace {

void nappMessageFreeCb(uint16_t /* eventType */, void *eventData) {
  chreHeapFree(eventData);
}

}  // namespace

ChreChannelOutputBase::ChreChannelOutputBase() : ChannelOutput("CHRE") {}

void ChreChannelOutputBase::setEndpointId(uint16_t endpointId) {
  mEndpointId = endpointId;
}

size_t ChreChannelOutputBase::MaximumTransmissionUnit() {
  return CHRE_MESSAGE_TO_HOST_MAX_SIZE;
}

void ChreNanoappChannelOutput::setNanoappEndpoint(uint32_t nanoappInstanceId) {
  CHRE_ASSERT(nanoappInstanceId <= UINT16_MAX);
  if (nanoappInstanceId <= UINT16_MAX) {
    mEndpointId = static_cast<uint16_t>(nanoappInstanceId);
  } else {
    mEndpointId = CHRE_HOST_ENDPOINT_UNSPECIFIED;
  }
}

pw::Status ChreNanoappChannelOutput::Send(std::span<const std::byte> buffer) {
  CHRE_ASSERT(mEndpointId != CHRE_HOST_ENDPOINT_UNSPECIFIED);
  pw::Status returnCode = PW_STATUS_OK;

  if (buffer.size() > 0) {
    auto *data = static_cast<ChrePigweedNanoappMessage *>(
        chreHeapAlloc(buffer.size() + sizeof(ChrePigweedNanoappMessage)));
    if (data == nullptr) {
      returnCode = PW_STATUS_RESOURCE_EXHAUSTED;
    } else {
      data->msgSize = buffer.size();
      memcpy(data->msg, buffer.data(), buffer.size());
      if (!chreSendEvent(PW_RPC_CHRE_NAPP_EVENT_TYPE, data, nappMessageFreeCb,
                         mEndpointId)) {
        returnCode = PW_STATUS_INVALID_ARGUMENT;
      }
    }
  }

  return returnCode;
}

void ChreHostChannelOutput::setHostEndpoint(uint16_t hostEndpoint) {
  setEndpointId(hostEndpoint);
}

pw::Status ChreHostChannelOutput::Send(std::span<const std::byte> buffer) {
  CHRE_ASSERT(mEndpointId != CHRE_HOST_ENDPOINT_UNSPECIFIED);
  pw::Status returnCode = PW_STATUS_OK;

  if (buffer.size() > 0) {
    uint8_t *data = memoryAlloc<uint8_t>(buffer.size());
    if (data == nullptr) {
      returnCode = PW_STATUS_RESOURCE_EXHAUSTED;
    } else {
      memcpy(data, buffer.data(), buffer.size());
      // TODO(b/210138227): Make this pass permissions too.
      if (!chreSendMessageWithPermissions(
              data, buffer.size(), PW_RPC_CHRE_HOST_MESSAGE_TYPE, mEndpointId,
              CHRE_MESSAGE_PERMISSION_NONE, heapFreeMessageCallback)) {
        returnCode = PW_STATUS_INVALID_ARGUMENT;
      }
    }
  }

  return returnCode;
}

}  // namespace chre

File: system/chre/java/utils/pigweed/src/com/google/android/chre/utils/pigweed/ChreChannelOutput.java
/*
 * Copyright (C) 2022 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.google.android.chre.utils.pigweed;

import android.hardware.location.ContextHubClient;
import android.hardware.location.ContextHubTransaction;
import android.hardware.location.NanoAppMessage;

import java.util.concurrent.atomic.AtomicBoolean;

import dev.pigweed.pw_rpc.Channel;
import dev.pigweed.pw_rpc.ChannelOutputException;

/**
 * Implements the Channel.Output interface of Pigweed RPC and provides a layer of abstraction on top
 * of Pigweed RPC to make it more friendly to use with CHRE APIs.
 */
public class ChreChannelOutput implements Channel.Output {
    /**
     * Random value chosen not too close to max value to try to avoid conflicts with other messages
     * in case pw rpc isn't the only way the client chooses to communicate.
     */
    public static final int PW_RPC_CHRE_MESSAGE_TYPE = Integer.MAX_VALUE - 10;

    // 1 denotes that a host endpoint is the client that created the channel.
    private static final int CHANNEL_ID_HOST_CLIENT = (1 << 16);

    private final ContextHubClient mClient;
    private final long mNanoappId;

    // Whether this output channel is no longer able to communicate with the nanoapp.
    private AtomicBoolean mAuthDenied = new AtomicBoolean(false);

    public ChreChannelOutput(ContextHubClient client, long nanoappId) {
        mClient = client;
        mNanoappId = nanoappId;
    }

    /**
     * This method MUST NOT be called directly from users of this class.
     */
    @Override
    public void send(byte[] packet) throws ChannelOutputException {
        NanoAppMessage message = NanoAppMessage.createMessageToNanoApp(
                mNanoappId, PW_RPC_CHRE_MESSAGE_TYPE, packet);
        if (mAuthDenied.get()
                || ContextHubTransaction.RESULT_SUCCESS != mClient.sendMessageToNanoApp(message)) {
            throw new ChannelOutputException();
        }
    }

    /**
     * @return Channel ID to use for all Channels that use this output to send
     *     messages to a nanoapp.
     */
    public int getChannelId() {
        return (CHANNEL_ID_HOST_CLIENT | mClient.getId());
    }

    /**
     * Used to indicate whether the particular nanoapp cannot be communicated
     * with any more (e.g. due to permissions loss).
     */
    void setAuthDenied(boolean denied) {
        mAuthDenied.set(denied);
    }
}
